pkg/store/pruning.go

```diff
@@ -0,0 +1,75 @@
+package store
+
+import (
+	"context"
+
+	ds "github.com/ipfs/go-datastore"
+
+	"github.com/rollkit/rollkit/pkg/config"
+	"github.com/rollkit/rollkit/types"
+)
+
+type DefaultPruningStore struct {
+	Store
+
+	Config config.PruningConfig
+}
+
+var _ PruningStore = &DefaultPruningStore{}
+
+// NewDefaultPruningStore returns default pruning store.
+func NewDefaultPruningStore(ds ds.Batching, config config.PruningConfig) PruningStore {
+	return &DefaultPruningStore{
+		Store:  &DefaultStore{db: ds},
+		Config: config,
+	}
+}
+
+// SaveBlockData adds block header and data to the store along with corresponding signature.
+// It also prunes the block data if needed.
+func (s *DefaultPruningStore) SaveBlockData(ctx context.Context, header *types.SignedHeader, data *types.Data, signature *types.Signature) error {
+	if err := s.PruneBlockData(ctx); err != nil {
+		return err
+	}
+
+	return s.Store.SaveBlockData(ctx, header, data, signature)
+}
+
+func (s *DefaultPruningStore) PruneBlockData(ctx context.Context) error {
+	var (
+		err error
+	)
+
+	// Skip if strategy is none.
+	if s.Config.Strategy == config.PruningConfigStrategyNone {
+		return nil
+	}
+
+	height, err := s.Height(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Skip if it's a correct interval or latest height is less or equal than number of blocks need to keep.
+	if height%s.Config.Interval != 0 || height < s.Config.KeepRecent {
+		return nil
+	}
+
+	// Must keep at least 2 blocks(while strategy is everything).
+	endHeight := height + 1 - s.Config.KeepRecent
+	startHeight := uint64(0)
+	if endHeight > s.Config.Interval {
+		startHeight = endHeight - s.Config.Interval
+	}
+
+	for i := startHeight; i < endHeight; i++ {
+		err = s.DeleteBlockData(ctx, i)
```

Comment:
pruning should be async, if we have it here it could end up blocking the main thread for an unknown amount of time.